validator/keymanager/remote-web3signer/keymanager.go

```diff
@@ -38,54 +47,235 @@ type SetupConfig struct {
+			log.WithField("path", km.keyFilePath).Warn("key file does not exist. please create a new one. the file should contain public keys in hex format 1 on each line.")
+		}
+	}
+
+	var ppk []string
+	// load key values
+	if cfg.PublicKeysURL != "" {
+		providedPublicKeys, err := km.client.GetPublicKeys(ctx, cfg.PublicKeysURL)
 		if err != nil {
 			erroredResponsesTotal.Inc()
-			return nil, errors.Wrap(err, fmt.Sprintf("could not get public keys from remote server url: %v", km.publicKeysURL))
+			return nil, errors.Wrap(err, fmt.Sprintf("could not get public keys from remote server url: %v", cfg.PublicKeysURL))
+		}
+		ppk = providedPublicKeys
+	} else if len(cfg.ProvidedPublicKeys) != 0 {
+		ppk = cfg.ProvidedPublicKeys
+	}
+
+	// use a map to remove duplicates
+	flagLoadedKeys := make(map[string][48]byte)
+
+	// Populate the map with existing keys
+	for _, key := range ppk {
+		decodedKey, err := hexutil.Decode(key)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not decode public key for web3signer")
+		}
+		if len(decodedKey) != fieldparams.BLSPubkeyLength {
+			return nil, fmt.Errorf("invalid public key length: expected %d, got %d", fieldparams.BLSPubkeyLength, len(decodedKey))
+		}
+		flagLoadedKeys[key] = bytesutil.ToBytes48(decodedKey)
+	}
+	if len(flagLoadedKeys) != 0 {
+		km.flagLoadedPublicKeys = maps.Values(flagLoadedKeys)
+	}
+
+	// load from file
+	if keyFileExists {
+		_, fileKeys, err := km.readKeyFile()
+		if err != nil {
+			return nil, errors.Wrap(err, "Could not read key file")
+		}
+		maps.Copy(fileKeys, flagLoadedKeys)
+		km.lock.Lock()
+		km.providedPublicKeys = maps.Values(fileKeys)
+		km.lock.Unlock()
+		// create a file watcher
+		go km.refreshRemoteKeysFromFileChanges(ctx)
+	} else {
+		km.lock.Lock()
+		km.providedPublicKeys = maps.Values(flagLoadedKeys)
+		km.lock.Unlock()
+	}
+
+	return km, nil
+}
+
+func (km *Keymanager) readKeyFile() ([][48]byte, map[string][48]byte, error) {
+	km.lock.RLock()
+	defer km.lock.RUnlock()
+
+	if km.keyFilePath == "" {
+		return nil, nil, errors.New("no key file path provided")
+	}
+	f, err := os.Open(filepath.Clean(km.keyFilePath))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not open remote signer public key file")
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.WithError(err).Error("could not close remote signer public key file")
+		}
+	}()
+	// Use a map to track and skip duplicate lines
+	seenKeys := make(map[string][48]byte)
+	scanner := bufio.NewScanner(f)
+	var keys [][48]byte
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		pubkeyLength := (fieldparams.BLSPubkeyLength * 2) + 2
+		if line == "" {
+			// skip empty line
+			continue
 		}
-		// makes sure that if the public keys are deleted the validator does not call URL again.
-		km.publicKeysUrlCalled = true
-		km.providedPublicKeys = providedPublicKeys
+		if len(line) != pubkeyLength {
+			log.Fatalf("remote signer key file: invalid public key: %s \n", line)
```

Comment:
Capital letter for logs.
Is the `\n` necessary?

Reply:
Also, it the pubkey is present in the file without the `0x` prefix, then this message is displayed:

```
[2024-05-31 17:21:10] FATAL remote signer key file: invalid public key: 80f408a41341416f79e3be605c5f1b960169bea5753b9b8181f7466b4fa0b13addb250c9d068af263f820c113536f9f7
```

which is quite confusing because this is a perfectly valid pubkey ... but without the `0x` prefix.
Would it be possible to handle both cases?
(As an ex node operator, it is something very useful IMO.)

Reply:
Also, if we write a key with bad characters inside, like:
```
0xz0f408a41341416f79e3be605c5f1b960169bea5753b9b8181f7466b4fa0b13addb250c9d068af263f820c113536f9f7
```

Then we got this error message:
```
[2024-05-31 17:26:07] FATAL client: Wallet is not ready error=could not initialize key manager: could not initialize web3signer keymanager: Could not read key file: invalid hex string
```

It could be very useful to actually print the faulty key in the error message.